### PR TITLE
add external related content (WP-741)

### PR DESCRIPTION
**Pull request by vsolovei-smartling**

### CODE REVIEW

File: inc/Smartling/ContentTypes/ExternalContentElementor.php [hunk 1]
```diff
@@ -133,36 +137,41 @@ public function alterContentFieldsForUpload(array $source): array
...
                     if (is_array($setting)) {
                         foreach ($setting as $id => $option) {
                             if (is_array($option)) {
-                                $options = array_filter($option, static function ($k) {
-                                    return strpos($k, '_') !== 0;
-                                }, ARRAY_FILTER_USE_KEY);
+                                foreach ($option as $optionKey => $optionValue) {
+                                    if (strpos($optionKey, '_') === 0) {
+                                        continue;
+                                    }
```
Comment (PavelLoparev) [inline inc/Smartling/ContentTypes/ExternalContentElementor.php:163]: Looks the same to me

File: inc/Smartling/ContentTypes/ExternalContentElementor.php [hunk 2]
```diff
@@ -197,11 +211,42 @@ public function getPluginPath(): string
...
+                break;
+            case 1:
+                $targetId = $targetSubmissions[0]->getTargetId();
+                if ($targetId !== 0) {
+                    return $targetId;
+                }
+                $this->getLogger()->info('Got 0 target attachment id for sourceId=' . $attachmentId);
+                break;
+            default:
+                $this->getLogger()->notice('Found more than one submissions while getting target attachmentId for sourceId=' . $attachmentId);
```
Comment (PavelLoparev) [inline inc/Smartling/ContentTypes/ExternalContentElementor.php:238]: Will be a strange situation when logs say found >1 submission but the actual exception is `EntityNotFoundException`. Maybe let's introduce a new exception for this case?

File: inc/Smartling/ContentTypes/ExternalContentElementor.php [hunk 3]
```diff
@@ -197,11 +211,42 @@ public function getPluginPath(): string
...
+            case 0:
+                $this->getLogger()->debug('No submissions found while getting target attachmentId for sourceId=' . $attachmentId);
+                break;
+            case 1:
+                $targetId = $targetSubmissions[0]->getTargetId();
+                if ($targetId !== 0) {
+                    return $targetId;
+                }
+                $this->getLogger()->info('Got 0 target attachment id for sourceId=' . $attachmentId);
+                break;
```
Comment (PavelLoparev) [inline inc/Smartling/ContentTypes/ExternalContentElementor.php:236]: `EntityNotFoundException` stands for submission entity or attachment? In this case, submission is found but no attachments - still `EntityNotFoundException`?

File: inc/Smartling/ContentTypes/ExternalContentManager.php [hunk 4]
```diff
@@ -50,6 +50,22 @@ public function getExternalContent(array $source, SubmissionEntity $submission,
...
+        foreach ($this->handlers as $handler) {
+            if ($handler->canHandle($this->pluginHelper)) {
+                $this->getLogger()->debug("Determined support for {$handler->getPluginId()}, will try to get related content");
+                try {
+                    return $handler->getRelatedContent($contentType, $id);
+                } catch (\Error $e) {
+                    $this->getLogger()->notice('HandlerName="' . $handler->getPluginId() . '" got exception while trying to get external related content: ' . $e->getMessage());
+                }
+            }
+        }
```
Comment (PavelLoparev) [inline inc/Smartling/ContentTypes/ExternalContentManager.php:64]: Is it possible that many handlers can handle related content? Seems like this implementation is `first wins`.

File: inc/Smartling/ContentTypes/ExternalContentElementor.php (hunk 1 above)
Comment (vsolovei-smartling) [reply inc/Smartling/ContentTypes/ExternalContentElementor.php:163]: Made consistent with the rest of the checks in file

File: inc/Smartling/ContentTypes/ExternalContentElementor.php (hunk 3 above)
Comment (vsolovei-smartling) [reply inc/Smartling/ContentTypes/ExternalContentElementor.php:236]: 0 target id here means that the content was not sent for translation

File: inc/Smartling/ContentTypes/ExternalContentManager.php (hunk 4 above)
Comment (vsolovei-smartling) [reply inc/Smartling/ContentTypes/ExternalContentManager.php]: Fixed, thanks!

File: inc/Smartling/ContentTypes/ExternalContentElementor.php (hunk 2 above)
Comment (vsolovei-smartling) [reply inc/Smartling/ContentTypes/ExternalContentElementor.php:237]: Refactored code not to use exceptions

File: inc/Smartling/ContentTypes/ExternalContentElementor.php (hunk 3 above)
Comment (PavelLoparev) [reply inc/Smartling/ContentTypes/ExternalContentElementor.php:235]: I mean you throw the same exception for two different situations:
1. submission is not found
2. submission is found but the attachment is not

**Review by PavelLoparev — APPROVED** [review]

**Review by dimitrystd — APPROVED** [review]

**Pull request merged**